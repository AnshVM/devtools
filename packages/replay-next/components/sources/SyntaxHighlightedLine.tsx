import { Suspense } from "react";

import { parse, parsedTokensToHtml } from "replay-next/src/suspense/SyntaxParsingCache";
import { ParsedToken } from "replay-next/src/suspense/SyntaxParsingCache";

import Loader from "../Loader";
import styles from "./SyntaxHighlightedLine.module.css";

type Props = { code: string; className?: string; fileExtension?: string; tokens?: ParsedToken[] };

export default function SyntaxHighlightedLine({
  className = "",
  code,
  fileExtension,
  tokens,
}: Props) {
  return (
    <Suspense fallback={<Loader />}>
      <SyntaxHighlightedLineSuspends
        className={className}
        code={code}
        fileExtension={fileExtension}
        tokens={tokens}
      />
    </Suspense>
  );
}

function SyntaxHighlightedLineSuspends({
  code,
  className = "",
  fileExtension = ".js",
  tokens,
}: Props) {
  tokens = tokens || (parse(code, fileExtension)?.[0] ?? []);
  const html = parsedTokensToHtml(tokens);

  return (
    <span
      className={`${className} ${styles.Code}`}
      dangerouslySetInnerHTML={{ __html: html }}
      title={code}
    />
  );
}
