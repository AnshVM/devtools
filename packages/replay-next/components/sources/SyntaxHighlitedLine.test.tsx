import renderer from 'react-test-renderer';
import SyntaxHighlightedLine from "./SyntaxHighlightedLine";

const SNIPPET = 'button.addEventListener("click", () => onClick(button, index + 1));';
const SNIPPET_TOKENS = [
  { columnIndex: 4, types: ["variableName"], value: "button" },
  { columnIndex: 10, types: ["operator"], value: "." },
  { columnIndex: 11, types: ["propertyName"], value: "addEventListener" },
  { columnIndex: 27, types: ["punctuation"], value: "(" },
  { columnIndex: 28, types: ["string"], value: '"click"' },
  { columnIndex: 35, types: ["punctuation"], value: "," },
  { columnIndex: 36, types: null, value: " " },
  { columnIndex: 37, types: ["punctuation"], value: "(" },
  { columnIndex: 38, types: ["punctuation"], value: ")" },
  { columnIndex: 39, types: null, value: " " },
  { columnIndex: 40, types: ["punctuation"], value: "=>" },
  { columnIndex: 42, types: null, value: " " },
  { columnIndex: 43, types: ["variableName"], value: "onClick" },
  { columnIndex: 50, types: ["punctuation"], value: "(" },
  { columnIndex: 51, types: ["variableName"], value: "button" },
  { columnIndex: 57, types: ["punctuation"], value: "," },
  { columnIndex: 58, types: null, value: " " },
  { columnIndex: 59, types: ["variableName"], value: "index" },
  { columnIndex: 64, types: null, value: " " },
  { columnIndex: 65, types: ["operator"], value: "+" },
  { columnIndex: 66, types: null, value: " " },
  { columnIndex: 67, types: ["number"], value: "1" },
  { columnIndex: 68, types: ["punctuation"], value: ")" },
  { columnIndex: 69, types: ["punctuation"], value: ")" },
  { columnIndex: 70, types: ["punctuation"], value: ";" },
];

describe("syntax-highlighting", () => {
  it("syntax highliting should work without passing tokens to SyntaxHighlitedLine", async () => {
    const tree = renderer.create(<SyntaxHighlightedLine code={SNIPPET} />).toJSON()
    expect(tree).toMatchSnapshot();
  });

  it("syntax highliting should work when tokens are passed to SyntaxHighlitedLine", async () => {
    const tree = renderer.create(<SyntaxHighlightedLine code={SNIPPET} tokens={SNIPPET_TOKENS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
