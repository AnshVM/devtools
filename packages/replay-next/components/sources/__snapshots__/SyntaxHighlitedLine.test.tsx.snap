// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syntax-highlighting syntax highliting should work when tokens are passed to SyntaxHighlitedLine 1`] = `
<span
  className=" Code"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<span class=\\"tok-variableName\\">button</span><span class=\\"tok-operator\\">.</span><span class=\\"tok-propertyName\\">addEventListener</span><span class=\\"tok-punctuation\\">(</span><span class=\\"tok-string\\">&quot;click&quot;</span><span class=\\"tok-punctuation\\">,</span><span class=\\"undefined\\"> </span><span class=\\"tok-punctuation\\">(</span><span class=\\"tok-punctuation\\">)</span><span class=\\"undefined\\"> </span><span class=\\"tok-punctuation\\">=&gt;</span><span class=\\"undefined\\"> </span><span class=\\"tok-variableName\\">onClick</span><span class=\\"tok-punctuation\\">(</span><span class=\\"tok-variableName\\">button</span><span class=\\"tok-punctuation\\">,</span><span class=\\"undefined\\"> </span><span class=\\"tok-variableName\\">index</span><span class=\\"undefined\\"> </span><span class=\\"tok-operator\\">+</span><span class=\\"undefined\\"> </span><span class=\\"tok-number\\">1</span><span class=\\"tok-punctuation\\">)</span><span class=\\"tok-punctuation\\">)</span><span class=\\"tok-punctuation\\">;</span>",
    }
  }
  title="button.addEventListener(\\"click\\", () => onClick(button, index + 1));"
/>
`;

exports[`syntax-highlighting syntax highliting should work without passing tokens to SyntaxHighlitedLine 1`] = `
<span
  className=" Code"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<span class=\\"tok-variableName\\">button</span><span class=\\"tok-operator\\">.</span><span class=\\"tok-propertyName\\">addEventListener</span><span class=\\"tok-punctuation\\">(</span><span class=\\"tok-string\\">&quot;click&quot;</span><span class=\\"tok-punctuation\\">,</span><span class=\\"undefined\\"> </span><span class=\\"tok-punctuation\\">(</span><span class=\\"tok-punctuation\\">)</span><span class=\\"undefined\\"> </span><span class=\\"tok-punctuation\\">=&gt;</span><span class=\\"undefined\\"> </span><span class=\\"tok-variableName\\">onClick</span><span class=\\"tok-punctuation\\">(</span><span class=\\"tok-variableName\\">button</span><span class=\\"tok-punctuation\\">,</span><span class=\\"undefined\\"> </span><span class=\\"tok-variableName\\">index</span><span class=\\"undefined\\"> </span><span class=\\"tok-operator\\">+</span><span class=\\"undefined\\"> </span><span class=\\"tok-number\\">1</span><span class=\\"tok-punctuation\\">)</span><span class=\\"tok-punctuation\\">)</span><span class=\\"tok-punctuation\\">;</span>",
    }
  }
  title="button.addEventListener(\\"click\\", () => onClick(button, index + 1));"
/>
`;
